pkg/storage/in_memory.go

```diff
@@ -0,0 +1,58 @@
+package storage
+
+import (
+	"container/list"
+	"sync"
+)
+
+// NewInMemoryCappedContainer create new in-memory capped container with capacity limit.
+func NewInMemoryCappedContainer(capacity int) CappedContainer {
+	return &inMemoryCappedContainer{
+		list:     list.New(),
+		capacity: capacity,
+	}
+}
+
+type inMemoryCappedContainer struct {
+	list     *list.List
+	capacity int
+	lock     sync.RWMutex
+}
+
+// Len returns current size of container
+func (h *inMemoryCappedContainer) Len() int {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+	return h.list.Len()
+}
+
+// Capacity returns maximum limit of the capped container
+func (h *inMemoryCappedContainer) Capacity() int {
+	return h.capacity
+}
+
+// Add adds new item to container
+func (h *inMemoryCappedContainer) Add(record interface{}) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	h.list.PushFront(record)
+
+	// Drop items exceeding capacity limit.
+	if h.list.Len() > h.capacity {
+		h.list.Remove(h.list.Back())
+	}
+}
+
+// Stream writes items to returned channel
+func (h *inMemoryCappedContainer) Stream() <-chan interface{} {
+	stream := make(chan interface{})
+	go func() {
+		h.lock.RLock()
+		defer h.lock.RUnlock()
```

Comment:
Locking just for `e.next()` could lead to inconsistencies reading the stream at the current time since it could change over the time.
But you are right about the locking if no one reads the output channel. This could be solved by returning buffered channel with capacity same as the actual length of the container. But this code is same as before I did not touch it at all. Not sure of it should be done in this PR but I'll try it out.